internal/chart.go

```diff
@@ -0,0 +1,257 @@
+package ledger
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+type ChartAccountRules struct{}
+
+type ChartAccount struct {
+	Metadata map[string]string
+	Rules    ChartAccountRules
+}
+
+type ChartSegment struct {
+	VariableSegment *ChartVariableSegment
+	FixedSegments   map[string]ChartSegment
+	Account         *ChartAccount
+}
+
+type ChartVariableSegment struct {
+	ChartSegment
+
+	Pattern string
+	Label   string
+}
+
+type ChartOfAccounts map[string]ChartSegment
+
+const SegmentRegex = "^\\$?[a-zA-Z0-9_-]+$"
+
+var Regexp = regexp.MustCompile(SegmentRegex)
+
+func ValidateSegment(addr string) bool {
+	return Regexp.Match([]byte(addr))
+}
+
+func (s *ChartOfAccounts) UnmarshalJSON(data []byte) error {
+	var segment map[string]json.RawMessage
+	if err := json.Unmarshal(data, &segment); err != nil {
+		return err
+	}
+	out := make(map[string]ChartSegment)
+	for key, value := range segment {
+		if !ValidateSegment(key) || key[0] == '$' || key[0] == '_' {
+			return fmt.Errorf("invalid segment name: %v", key)
+		}
+		var seg ChartSegment
+		err := seg.UnmarshalJSON(value)
+		if err != nil {
+			return fmt.Errorf("invalid segment `%v`: %v", key, err)
+		}
+		out[key] = seg
+	}
+	*s = out
+	return nil
+}
+func (s *ChartSegment) UnmarshalJSON(data []byte) error {
+	var segment map[string]json.RawMessage
+	if err := json.Unmarshal(data, &segment); err != nil {
+		return err
+	}
+	isLeaf := true
+	var isAccount bool
+	var account ChartAccount
+	var fixedSegments map[string]ChartSegment
+	var variableSegment *ChartVariableSegment
+	keys := []string{}
+	for key := range segment {
+		keys = append(keys, key)
+	}
+	for _, key := range keys {
+		value := segment[key]
+		isSubsegment := key[0] != '_'
```

Comment:
So, we are denying any segment starting with '_' ?
What about existing users which would have already use that?